src/storage/src/source/mod.rs

```diff
@@ -999,38 +1042,40 @@ where
             // Record operator has been scheduled
             source_metrics.operator_scheduled_counter.inc();
 
-            let mut source_state = (SourceStatus::Alive, MessageProcessing::Active);
-            while let (_, MessageProcessing::Active) = source_state {
-                source_state = match source_reader.get_next_message() {
-                    Ok(NextMessage::Ready(message)) => {
+            let mut context = Context::from_waker(&waker);
+            let mut source_status = SourceStatus::Alive;
+            while let Poll::Ready(item) = source_reader.as_mut().poll_next(&mut context) {
```

Comment:
when we switch to this, we will be moving cpu-bound work like pg row-decoding off of tokio worker threads and into timely threads, which is a big perf-consistency win, but we should document that async io is still subtle to use (we are using the timely scheduler instead of the tokio one), and that things that require a tokio runtime will need to enter it in their async fn

Reply:
I don't think you have to enter the tokio runtime to use tokio io in timely because that only uses the reactor part of tokio which keeps around arbitrary wakers